### PR TITLE
Implement formula evaluator

### PR DESCRIPTION
**Pull request by nsmith-**
Following https://github.com/nsmith-/correctionlib/discussions/17#discussioncomment-315714

### CODE REVIEW

File: include/correction.h [hunk 1]
```diff
@@ -35,6 +36,7 @@ typedef std::variant<double, Binning, MultiBinning, Category, Formula> Content;
 class Formula {
   public:
     enum class ParserType {TFormula, numexpr};
+    static bool eager_compilation; // true by default
```
Comment (Dr15Jones) [inline include/correction.h:39]: Not thread safe

File: include/correction.h [hunk 2]
```diff
@@ -43,8 +45,40 @@ class Formula {
   private:
     std::string expression_;
     ParserType type_;
-    std::vector<int> parameterIdx_;
-    mutable std::unique_ptr<TFormula> evaluator_;
+    std::vector<int> variableIdx_;
+
+    static std::map<ParserType, peg::parser> parsers_;
+    static std::mutex parsers_mutex_; // could be one per parser, but this is good enough
```
Comment (Dr15Jones) [inline include/correction.h:51]: Have you done large scale threading tests to see how this affect performance?

File: src/correction.cc [hunk 3]
```diff
@@ -45,25 +96,186 @@ Formula::Formula(const rapidjson::Value& json) :
...
+      throw std::runtime_error(
+        "Failed to parse Formula expression at position " + std::to_string(pos) + ":\n"
+        + expression_ + "\n"
+        + std::string(pos, ' ') + "^\n"
+        + msg
+      );
     }
+    peg_ast = parser.optimize_ast(peg_ast);
+  }
+  ast_ = std::make_unique<Ast>(translate_ast(*peg_ast));
```
Comment (Dr15Jones) [inline src/correction.cc:134]: Do you know or sure that using peg_ast at this point is thread safe?

File: include/correction.h (hunk 1 above)
Comment (nsmith-) [reply include/correction.h:39]: Thanks for the review! Indeed I was thinking about this, but my reasoning is that whatever state this is in when the Formula is constructed decides the eagerness of that Formula object, and regardless of its value, the Formula will compile itself eventually. Basically some race condition might mean that some Formulas will be lazy when they were not desired, but the end result is harmless.

File: include/correction.h (hunk 2 above)
Comment (nsmith-) [reply include/correction.h:51]: No, but I hope the number of Formulas to compile is << the number of evaluations, so such contention is rare. Unless someone is doing something disgusting like re-loading the data each event...

File: include/correction.h (hunk 1 above)
Comment (Dr15Jones) [reply include/correction.h:39]: That is against CMS threading policy. The C++ standard says their are not guarnteed 'safe' race conditions and everything else is undefined behavior. This should become a std::atomic if you want this behavior. Better yet, remove the static and make laziness a parameter passed to the constructor.

File: src/correction.cc (hunk 3 above)
Comment (nsmith-) [reply src/correction.cc:134]: translate_ast takes a const reference, dunno if that's a guarantee. At least it appears to be a data container only.

File: include/correction.h (hunk 1 above)
Comment (nsmith-) [reply include/correction.h:39]: Ok, I mean it's two states, but undefined behavior is bad. I'll toss an atomic wrapper around it. The constructor is meant to take the data format, which shouldn't be concerned with evaluation.

File: src/correction.cc [hunk 4]
```diff
@@ -45,25 +96,186 @@ Formula::Formula(const rapidjson::Value& json) :
...
+    };
+  }
+  throw std::runtime_error("Unrecognized AST node");
+}
+
+double Formula::evaluate(const std::vector<Variable>& inputs, const std::vector<Variable::Type>& values) const {
+  std::vector<double> variables;
+  variables.reserve(variableIdx_.size());
+  for ( auto idx : variableIdx_ ) { variables.push_back(std::get<double>(values[idx])); }
+  if ( ! ast_ ) build_ast();
```
Comment (Dr15Jones) [inline src/correction.cc:243]: This makes Formula::evaluate non-reentrant and therefore not thread safe.

File: src/correction.cc [hunk 5]
```diff
@@ -45,25 +96,186 @@ Formula::Formula(const rapidjson::Value& json) :
...
+    return {
+      Ast::NodeType::Expression,
+      ast.nodes[1]->token[0],
+      {translate_ast(*ast.nodes[0]), translate_ast(*ast.nodes[2])}
+    };
+  }
+  throw std::runtime_error("Unrecognized AST node");
+}
+
+double Formula::evaluate(const std::vector<Variable>& inputs, const std::vector<Variable::Type>& values) const {
```
Comment (Dr15Jones) [inline src/correction.cc:239]: inputs does not appear to be used.

File: include/correction.h [hunk 6]
```diff
@@ -43,8 +45,40 @@ class Formula {
...
+        char, // unary / binary op
+        UnaryFcn,
+        BinaryFcn
+      > NodeData;
+
+      NodeType nodetype;
+      NodeData data;
+      std::vector<Ast> children;
+    };
+    mutable std::unique_ptr<Ast> ast_;
```
Comment (Dr15Jones) [inline include/correction.h:78]: using mutable is a pretty clear sign of non-thread safety.

File: src/correction.cc (hunk 4 above)
Comment (nsmith-) [reply src/correction.cc:243]: Shoot. Digging for idioms, I guess what I want is a thread safe lazy initialization. Seem `std::call_once` fits the bill?

File: src/correction.cc (hunk 3 above)
Comment (Dr15Jones) [reply src/correction.cc:134]: Looking at the peg code, I'm having a very hard time being able to prove it generates stuff that is reentrant. Lots of holding pointers, having mutating methods (e.g. the visitor pattern they use allows mutating the items in the node) as well as the use of `mutable`.

File: src/correction.cc (hunk 4 above)
Comment (Dr15Jones) [reply src/correction.cc:243]: The absolute best is to just do the ast building in the constructor. What is it about the usage pattern which is driving you not to do that? In general, delayed initialization in a threaded context is a very bad thing to do.

File: include/correction.h (hunk 2 above)
Comment (nsmith-) [reply include/correction.h:51]: Further, the number of parsers is very small (currently 1)

Comment (nsmith-) [reply include/correction.h:51]: Actually the map is more or less pointless: if I wrapped the peg::parser with a shim class I could construct that as a static const member.

File: src/correction.cc (hunk 3 above)
Comment (nsmith-) [reply src/correction.cc:134]: Ok, I'll move it inside the locked scope. This was pretty clear premature optimization

File: src/correction.cc (hunk 4 above)
Comment (nsmith-) [reply src/correction.cc:243]: There are sets of B-tagging scale factors that are a binned lookup into 18k (!!) formulas. Usually the user accesses only a portion of the phase space, so not parsing every formula can save time. Dramatically so for TFormula: using https://github.com/cms-sw/cmssw/blob/master/CondTools/BTau/interface/BTagCalibrationReader.h on the csv equivalent to the example file in this repo takes 5 minutes just to construct. That said, this PEG parser is really fast and can compile all 18k in about 2s so the motivation behind delayed init is reduced. Further, these formulas follow a regular pattern and I think if BTV adopted the JME solution of replacing float literals with parameters then there would only be a handful of formulas to compile.

**Pull request merged**